### PR TITLE
[Snyk] Security upgrade keras from 2.11.0 to 3.14.0

### PR DESCRIPTION
**Pull request by jhamot**
snyk-top-banner

#### Snyk changed the following file(s):

- `tf2/requirements.txt`

<details>
<summary>⚠️ <b>Warning</b></summary>

```
tensorflow 2.11.0 requires tensorflow-io-gcs-filesystem, which is not installed.
tensorflow 2.11.0 requires keras, which is not installed.
tensorflow-metadata 1.12.0 has requirement absl-py<2.0.0,>=0.9, but you have absl-py 2.1.0.
tensorboard 2.11.2 has requirement setuptools>=41.0.0, but you have setuptools 40.5.0.
googleapis-common-protos 1.73.0 has requirement protobuf!=4.21.1,!=4.21.2,!=4.21.3,!=4.21.4,!=4.21.5,<7.0.0,>=3.20.2, but you have protobuf 3.19.6.
```

</details>

#### Breaking Change Risk

Merge Risk: High

> **Notice:** This assessment is enhanced by AI.

---

> [!IMPORTANT]
>
> - Check the changes in this PR to ensure they won't cause issues with your project.
> - Max score is 1000. Note that the real score may have changed since the PR was raised.
> - This PR was automatically created by Snyk using the credentials of a real user.
> - Some vulnerabilities couldn't be fully fixed and so Snyk will still find them when the project is tested again. This may be because the vulnerability existed within more than one direct dependency, but not all of the affected dependencies could be upgraded.

---

**Note:** _You are seeing this because you or someone else with access to this repository has authorized Snyk to open fix PRs._

For more information: 
🧐 [View latest project report](https://app.snyk.io/org/lucasfilm/project/c1bd5c26-cdb9-4fef-9b86-a38cffefa71b?utm_source&#x3D;github&amp;utm_medium&#x3D;referral&amp;page&#x3D;fix-pr)
📜 [Customise PR templates](https://docs.snyk.io/scan-using-snyk/pull-requests/snyk-fix-pull-or-merge-requests/customize-pr-templates?utm_source=github&utm_content=fix-pr-template)
🛠 [Adjust project settings](https://app.snyk.io/org/lucasfilm/project/c1bd5c26-cdb9-4fef-9b86-a38cffefa71b?utm_source&#x3D;github&amp;utm_medium&#x3D;referral&amp;page&#x3D;fix-pr/settings)
📚 [Read about Snyk's upgrade logic](https://docs.snyk.io/scan-with-snyk/snyk-open-source/manage-vulnerabilities/upgrade-package-versions-to-fix-vulnerabilities?utm_source=github&utm_content=fix-pr-template)

---

**Learn how to fix vulnerabilities with free interactive lessons:**

🦉 [Deserialization of Untrusted Data](https://learn.snyk.io/lesson/insecure-deserialization/?loc&#x3D;fix-pr)

[//]: # 'snyk:metadata:{"breakingChangeRiskLevel":"high","FF_showPullRequestBreakingChanges":false,"FF_showPullRequestBreakingChangesWebSearch":true,"customTemplate":{"variablesUsed":[],"fieldsUsed":[]},"dependencies":[{"name":"keras","from":"2.11.0","to":"3.14.0"}],"env":"prod","issuesToFix":["SNYK-PYTHON-KERAS-16032293"],"prId":"b0c7e2c6-cd33-403e-b6c6-dbc34ca3a5a0","prPublicId":"b0c7e2c6-cd33-403e-b6c6-dbc34ca3a5a0","packageManager":"pip","priorityScoreList":[234],"projectPublicId":"c1bd5c26-cdb9-4fef-9b86-a38cffefa71b","projectUrl":"https://app.snyk.io/org/lucasfilm/project/c1bd5c26-cdb9-4fef-9b86-a38cffefa71b?utm_source=github&utm_medium=referral&page=fix-pr","prType":"fix","templateFieldSources":{"branchName":"default","commitMessage":"default","description":"default","title":"default"},"templateVariants":["updated-fix-title","pr-warning-shown","priorityScore"],"type":"auto","upgrade":[],"vulns":["SNYK-PYTHON-KERAS-16032293"],"patch":[],"isBreakingChange":false,"remediationStrategy":"vuln"}'

### CODE REVIEW

**Comment by jhamot** [comment]
Merge Risk: High

The upgrade from Keras 2.11.0 to 3.14.0 is a major rewrite with significant architectural changes and requires code modifications.  
  
**Key Breaking Changes:**  
  
*   **Multi-Backend Architecture**: Keras 3 is no longer just a TensorFlow frontend. It has been redesigned to support multiple backends, including TensorFlow, JAX, and PyTorch. Your existing code, which likely uses `tf.keras`, will need to be adapted.  
*   **Mandatory Import Changes**: All imports must be updated. For example, `from tensorflow.keras import layers` must be changed to `from keras import layers`.  
*   **API Removals**: Several deprecated methods from Keras 2 have been removed, including:  
    *   `predict_proba()` and `predict_classes()`  
    *   `model_to_yaml()` and `model_from_yaml()`  
*   **Backend-Specific Code**: To make code truly portable across backends, any direct use of TensorFlow APIs (e.g., `tf.some_function`) must be replaced with the backend-agnostic `keras.ops` equivalent (e.g., `keras.ops.some_function`).  
  
**Recommendation:**  
This is a high-impact upgrade. Developers must review their codebase to update all import statements and replace removed APIs. For projects that are not actively developed but need to keep running, an alternative is to switch to the `tf_keras` package, which is a maintained version of Keras 2.  
  
Source: [Keras 3 Migration Guide](https://keras.io/guides/migrating_to_keras_3/)

> **Notice 🤖:** This content was augmented using **artificial intelligence**. AI-generated content **may** contain **errors** and should be **reviewed** for accuracy before use.
